Makefile

```diff
@@ -1,323 +1,442 @@
+# architectures. (i.e. make docker-buildx IMG=myregistry/mypoperator:0.0.1). To use this option you need to:
+# - be able to use docker buildx. More info: https://docs.docker.com/build/buildx/
+# - have enabled BuildKit. More info: https://docs.docker.com/develop/develop-images/build_enhancements/
+# - be able to push the image to your registry (i.e. if you do not set a valid value via IMG=<myregistry/image:<tag>> then the export will fail)
+# To adequately provide solutions that are compatible with multiple platforms, you should consider using this option.
+PLATFORMS ?= linux/arm64,linux/amd64,linux/s390x,linux/ppc64le
+.PHONY: docker-buildx
+docker-buildx: ## Build and push docker image for the manager for cross-platform support
+	# copy existing Dockerfile and insert --platform=${BUILDPLATFORM} into Dockerfile.cross, and preserve the original Dockerfile
+	sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' Dockerfile > Dockerfile.cross
+	- $(CONTAINER_TOOL) buildx create --name project-v3-builder
+	$(CONTAINER_TOOL) buildx use project-v3-builder
+	- $(CONTAINER_TOOL) buildx build --push --platform=$(PLATFORMS) --tag ${OPERATOR_IMG} -f Dockerfile.cross .
+	- $(CONTAINER_TOOL) buildx rm project-v3-builder
+	rm Dockerfile.cross
+
+##@ Deployment
+
+ifndef ignore-not-found
+  ignore-not-found = true
+endif
 
-.PHONY: help
-help:
-	@grep -E '^[a-zA-Z/0-9_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+.PHONY: install
+install: uninstall manifests kustomize ## Install CRDs into the K8s cluster specified in ~/.kube/config.
+	$(KUSTOMIZE) build config/crd | $(KUBECTL) apply -f -
 
-.PHONY: clean
-clean:  ## Remove build artifacts
-	rm -rf $(OUTPUT_DIR) bin
+.PHONY: uninstall
+uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
+	$(KUSTOMIZE) build config/crd | $(KUBECTL) delete --ignore-not-found=$(ignore-not-found) -f -
 
-.PHONY: clean-deploy
-clean-deploy: ## Uninstall VPA from current cluster
-	${KUBECTL} delete mutatingwebhookconfigurations vpa-webhook-config || true
-	${KUBECTL} delete ns openshift-vertical-pod-autoscaler || true
-	${KUBECTL} delete crd verticalpodautoscalercheckpoints.autoscaling.k8s.io verticalpodautoscalercontrollers.autoscaling.openshift.io verticalpodautoscalers.autoscaling.k8s.io || true
+# 	This is a HACK. The environment variable that is being replaced must be in the same line referenced in VPA_REPLACE_LINE.
+VPA_REPLACE_LINE := 14
+.PHONY: predeploy
+predeploy: ## Setup configuration for the operator before deployment.
+	cd config/manager && $(KUSTOMIZE) edit set image quay.io/openshift/origin-vertical-pod-autoscaler-operator=$(OPERATOR_IMG)
+	sed -i '$(VPA_REPLACE_LINE)s|value:.*|value: $(OPERAND_IMG)|' $(VPA_OPERAND_REPLACE_FILE)
```

Comment:
Hahaha so wait:
- we ultimately need to overide the operand value so we make it an ENV variable 
- but in order for that to happen, we need kustomize to rewrite `manager.yaml` with that ENV variable
- but in order for that to happen, we need to add a patch to the `kustomization.yaml` patch list containing that ENV variable
- but the patch will be wrong unless we replace the value of that ENV variable in the kustomization patch using sed ?

Would this be less fragile if it the sed replacement was anchored on matching the value directly instead of the line number? Or is there a reason we can't do that? 

Or....maybe is there a way we could do it with `kustomize edit remove patch` / `kustomize edit add patch`  or something?   I understand how hacky/unwieldy kustomize is, theory vs practice, etc so the answer might be "yes but no"  :stuck_out_tongue: )

Reply:
Yes, unfortunately that's what is happening :smiling_face_with_tear:. I don't remember why I didn't just sed directly and skip the 10 other steps, but I think I tunneled on trying to solve this with Kustomize. It's probably doable with a patch file, without sed now that i think about it. I'll try it out.

Reply:
Okay, I think I was able to make it cleaner with `kustomize edit remove/add patch` and `yq`.

Reply:
Makes sense! Thanks!  

I think it's totally fine how you're downloading `yq` right now, but it is a go program, so you could probably skip all that if you wanted to and do like a `go install github.com/mikefarah/yq/v4@latest` ( this is not "plz change it" this is "fyi" :smile: )